File: python/mirage/commands/builtin/grep_binary.py
import re
from collections import deque
from collections.abc import AsyncIterator
from dataclasses import dataclass, replace

from mirage.commands.builtin.grep_select import WalkFilters
from mirage.io.async_line_iterator import AsyncLineIterator
from mirage.io.stream import close_quietly
from mirage.io.types import IOResult, materialize


@dataclass(frozen=True, slots=True)
class GrepFlags:
    """Parsed grep flags (TS FlagSet parity); the complete set grep honors."""
    ignore_case: bool
    invert: bool
    line_numbers: bool
    count_only: bool
    files_only: bool
    whole_word: bool
    fixed_string: bool
    basic_regexp: bool
    only_matching: bool
    quiet: bool
    recursive: bool
    with_filename: bool
    no_filename: bool
    max_count: int | None
    after_context: int
    before_context: int
    binary_mode: str
    filters: WalkFilters


class BinaryInput:

    def __init__(self, mode: str) -> None:
        self.mode = mode
        self.nul = False

    async def read(self, source: AsyncIterator[bytes]) -> AsyncIterator[bytes]:
        async for chunk in binary_blocks(source):
            if self.mode != "text" and b"\0" in chunk:
                self.nul = True
                if self.mode == "without-match":
                    return
            yield chunk.replace(b"\0", b"\n") if self.nul else chunk


def valid_utf8(data: bytes) -> bool:
    try:
        data.decode("utf-8")
        return True
    except UnicodeDecodeError:
        return False


def output_line(raw: bytes, number: int, selected: bool, path: str,
                show_filename: bool, f: GrepFlags) -> bytes:
    separator = b":" if selected else b"-"
    prefix = path.encode() + separator if show_filename else b""
    if f.line_numbers:
        prefix += str(number).encode() + separator
    return prefix + raw + b"\n"


async def grep_input(source: AsyncIterator[bytes], pat: re.Pattern[str],
                     f: GrepFlags, path: str, show_filename: bool,
                     io: IOResult) -> AsyncIterator[bytes]:
    io.exit_code = 1
    pat = utf8_pattern(pat)
    binary = BinaryInput(f.binary_mode)
    count = 0
    notified = False
    previous: deque[tuple[int, bytes]] = deque(maxlen=f.before_context)
    last_printed = 0
    after_until = 0
    has_context = bool(f.after_context
                       or f.before_context) and not f.only_matching
    if f.max_count == 0:
        if f.count_only and not (f.quiet or f.files_only):
            yield output_line(b"0", 0, True, path, show_filename,
                              replace(f, line_numbers=False))
        return
    number = 0
    input_stream = binary.read(source)
    try:
        async for raw in AsyncLineIterator(input_stream):
            if binary.nul and f.binary_mode == "without-match":
                break
            number += 1
            # Surrogate escapes preserve raw bytes under -a and -ao.
            line = raw.decode("utf-8", errors="surrogateescape")
            hit = bool(pat.search(line)) != f.invert
            if f.max_count is not None and count >= f.max_count:
                hit = False
            if hit:
                count += 1
                io.exit_code = 0
                if f.quiet:
                    return
                if f.files_only:
                    yield path.encode() + b"\n"
                    return
            if f.count_only:
                if f.max_count is not None and count >= f.max_count:
                    break
                continue
            chunks: list[bytes] = []
            if hit:
                if f.only_matching:
                    if not f.invert:
                        chunks = [
                            output_line(
                                m.group().encode("utf-8",
                                                 errors="surrogateescape"),
                                number, True, path, show_filename, f)
                            for m in pat.finditer(line) if m.group()
                        ]
                else:
                    if has_context:
                        pending = [(n, data) for n, data in previous
                                   if n > last_printed]
                        first = pending[0][0] if pending else number
                        if last_printed and first > last_printed + 1:
                            chunks.append(b"--\n")
                        chunks.extend(
                            output_line(data, n, False, path, show_filename, f)
                            for n, data in pending)
                    chunks.append(
                        output_line(raw, number, True, path, show_filename, f))
                    last_printed = number
                    after_until = number + f.after_context
            elif has_context and number <= after_until:
                chunks.append(
                    output_line(raw, number, False, path, show_filename, f))
                last_printed = number
            for chunk in chunks:
                if f.binary_mode != "text" and (binary.nul
                                                or not valid_utf8(chunk)):
                    if f.binary_mode == "binary" and not notified:
                        io.stderr = (await materialize(
                            io.stderr
                        )) + f"grep: {path}: binary file matches\n".encode()
                        notified = True
                    continue
                yield chunk
            if binary.nul and count and f.binary_mode == "binary":
                return
            previous.append((number, raw))
            if (f.max_count is not None and count >= f.max_count
                    and number >= after_until):
                break
    finally:
        await close_quietly(input_stream)
        await close_quietly(source)

    # Detection can end input without yielding another line.
    if binary.nul and f.binary_mode == "without-match":
        count = 0
        io.exit_code = 1

    if f.count_only and not (f.quiet or f.files_only):
        yield (f"{path}:"
               if show_filename else "").encode() + f"{count}\n".encode()


async def binary_blocks(source: AsyncIterator[bytes]) -> AsyncIterator[bytes]:
    # Inspect available bytes without pulling ahead from a remote row stream.
    # Like GNU, detection can suppress only input not already emitted.
    async for chunk in source:
        for offset in range(0, len(chunk), 32768):
            yield chunk[offset:offset + 32768]


def utf8_pattern(pat: re.Pattern[str]) -> re.Pattern[str]:
    parts: list[str] = []
    escaped = False
    in_class = False
    for char in pat.pattern:
        if escaped:
            parts.append(char)
            escaped = False
        elif char == "\\":
            parts.append(char)
            escaped = True
        elif char == "[":
            parts.append(char)
            in_class = True
        elif char == "]":
            parts.append(char)
            in_class = False
        elif char == "." and not in_class:
            parts.append(r"[^\n\udc80-\udcff]")
        else:
            parts.append(char)
    return re.compile("".join(parts), pat.flags)

File: typescript/packages/core/src/commands/builtin/grep_binary.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { byteChar, encodeText } from '../../shell/bytes.ts'
import { closeQuietly } from '../../io/stream.ts'
import { AsyncLineIterator } from '../../io/async_line_iterator.ts'
import { concat } from '../../io/cachable_iterator.ts'
import type { IOResult } from '../../io/types.ts'
import type { WalkFilters } from './grep_select.ts'

const ENC = new TextEncoder()
const DEC = new TextDecoder('utf-8', { ignoreBOM: true })

export interface FlagSet {
  filters: WalkFilters
  binaryMode: string
  recursive: boolean
  ignoreCase: boolean
  invert: boolean
  lineNumbers: boolean
  countOnly: boolean
  filesOnly: boolean
  wholeWord: boolean
  fixedString: boolean
  basicRegexp: boolean
  onlyMatching: boolean
  maxCount: number | null
  quiet: boolean
  withFilename: boolean
  noFilename: boolean
  afterContext: number
  beforeContext: number
}

// GNU grep's INITIAL_BUFSIZE: the window it examines before printing from it.
export const PROBE_BLOCK_BYTES = 96 * 1024

export class BinaryInput {
  nul = false
  constructor(readonly mode: string) {}

  /**
   * Regroup the transport's chunks into GNU-sized probe blocks. GNU reads a
   * whole buffer before printing from it, so a NUL anywhere in the window
   * suppresses the lines ahead of it however the transport chunked them. A
   * NUL is acted on the moment it arrives, and nothing past the block the
   * scanner asked for is read.
   */
  async *read(source: AsyncIterable<Uint8Array>): AsyncIterable<Uint8Array> {
    let pending: Uint8Array[] = []
    let size = 0
    for await (const chunk of source) {
      pending.push(chunk)
      size += chunk.length
      if (size < PROBE_BLOCK_BYTES) {
        // Still inside one block, so a NUL here is that block's.
        if (this.stops(chunk)) return
        continue
      }
      const data = concat(pending)
      const whole = data.length - (data.length % PROBE_BLOCK_BYTES)
      for (let offset = 0; offset < whole; offset += PROBE_BLOCK_BYTES) {
        const block = data.subarray(offset, offset + PROBE_BLOCK_BYTES)
        if (this.stops(block)) return
        yield this.deliver(block)
      }
      const rest = data.subarray(whole)
      pending = rest.length > 0 ? [rest] : []
      size = rest.length
      if (rest.length > 0 && this.stops(rest)) return
    }
    if (size > 0) yield this.deliver(concat(pending))
  }

  // Note a NUL in data, which all belongs to the block being probed; true
  // when without-match must stop reading.
  private stops(data: Uint8Array): boolean {
    if (this.mode !== 'text' && !this.nul && data.includes(0)) this.nul = true
    return this.nul && this.mode === 'without-match'
  }

  private deliver(block: Uint8Array): Uint8Array {
    return this.nul ? block.map((byte) => (byte === 0 ? 10 : byte)) : block
  }
}

export function validUtf8(data: Uint8Array): boolean {
  try {
    new TextDecoder('utf-8', { fatal: true }).decode(data)
    return true
  } catch (error) {
    if (!(error instanceof TypeError)) throw error
    return false
  }
}

function outputLine(
  raw: Uint8Array,
  number: number,
  selected: boolean,
  path: string,
  showFilename: boolean,
  f: FlagSet,
): Uint8Array {
  const separator = selected ? ':' : '-'
  const prefix = ENC.encode(
    (showFilename ? path + separator : '') + (f.lineNumbers ? String(number) + separator : ''),
  )
  const out = new Uint8Array(prefix.length + raw.length + 1)
  out.set(prefix)
  out.set(raw, prefix.length)
  out[out.length - 1] = 10
  return out
}

export async function* grepInput(
  source: AsyncIterable<Uint8Array>,
  pat: RegExp,
  f: FlagSet,
  path: string,
  showFilename: boolean,
  io: IOResult,
  // Whether an earlier input already printed lines; GNU then opens this
  // input's first context group with the separator, as it does between
  // groups within one input.
  afterOutput = false,
): AsyncIterable<Uint8Array> {
  io.exitCode = 1
  pat = utf8Pattern(pat)
  const binary = new BinaryInput(f.binaryMode)
  let count = 0
  let notified = false
  const previous: [number, Uint8Array][] = []
  let lastPrinted = 0
  let afterUntil = 0
  const hasContext = (f.afterContext > 0 || f.beforeContext > 0) && !f.onlyMatching
  if (f.maxCount === 0) {
    if (f.countOnly && !(f.quiet || f.filesOnly))
      yield ENC.encode((showFilename ? path + ':' : '') + '0\n')
    return
  }
  let number = 0
  const input = binary.read(source)
  try {
    for await (const raw of new AsyncLineIterator(input)) {
      if (binary.nul && f.binaryMode === 'without-match') break
      number += 1
      const line = decodeLine(raw)
      let hit = pat.test(line) !== f.invert
      if (f.maxCount !== null && count >= f.maxCount) hit = false
      if (hit) {
        count += 1
        io.exitCode = 0
        if (f.quiet) return
        if (f.filesOnly) {
          yield ENC.encode(path + '\n')
          return
        }
      }
      if (f.countOnly) {
        if (f.maxCount !== null && count >= f.maxCount) break
        continue
      }
      const chunks: Uint8Array[] = []
      if (hit) {
        if (f.onlyMatching) {
          if (!f.invert) {
            const re = new RegExp(pat.source, pat.flags.includes('g') ? pat.flags : pat.flags + 'g')
            for (const m of line.matchAll(re)) {
              if (m[0] !== '')
                chunks.push(outputLine(encodeText(m[0]), number, true, path, showFilename, f))
            }
          }
        } else {
          if (hasContext) {
            const pending = previous.filter(([n]) => n > lastPrinted)
            const first = pending[0]?.[0] ?? number
            if ((lastPrinted && first > lastPrinted + 1) || (!lastPrinted && afterOutput))
              chunks.push(ENC.encode('--\n'))
            for (const [n, data] of pending)
              chunks.push(outputLine(data, n, false, path, showFilename, f))
          }
          chunks.push(outputLine(raw, number, true, path, showFilename, f))
          lastPrinted = number
          afterUntil = number + f.afterContext
        }
      } else if (hasContext && number <= afterUntil) {
        chunks.push(outputLine(raw, number, false, path, showFilename, f))
        lastPrinted = number
      }
      for (const chunk of chunks) {
        if (f.binaryMode !== 'text' && (binary.nul || !validUtf8(chunk))) {
          if (f.binaryMode === 'binary' && !notified) {
            const old = io.stderr instanceof Uint8Array ? io.stderr : new Uint8Array()
            const notice = ENC.encode(`grep: ${path}: binary file matches\n`)
            const err = new Uint8Array(old.length + notice.length)
            err.set(old)
            err.set(notice, old.length)
            io.stderr = err
            notified = true
          }
          continue
        }
        yield chunk
      }
      if (binary.nul && count && f.binaryMode === 'binary') return
      previous.push([number, raw])
      if (previous.length > f.beforeContext) previous.shift()
      if (f.maxCount !== null && count >= f.maxCount && number >= afterUntil) break
    }
  } finally {
    await closeQuietly(input)
    await closeQuietly(source)
  }

  // Detection can end input without yielding another line.
  if (binary.nul && f.binaryMode === 'without-match') {
    count = 0
    io.exitCode = 1
  }

  if (f.countOnly && !(f.quiet || f.filesOnly))
    yield ENC.encode((showFilename ? path + ':' : '') + String(count) + '\n')
}

function decodeLine(raw: Uint8Array): string {
  if (validUtf8(raw)) return DEC.decode(raw)
  let text = ''
  for (let i = 0; i < raw.length; ) {
    const byte = raw[i]
    if (byte === undefined) break
    const width = byte < 0x80 ? 1 : byte < 0xe0 ? 2 : byte < 0xf0 ? 3 : 4
    const part = raw.subarray(i, i + width)
    if (part.length === width && validUtf8(part)) {
      text += DEC.decode(part)
      i += width
    } else {
      text += byteChar(byte)
      i += 1
    }
  }
  return text
}

function utf8Pattern(pat: RegExp): RegExp {
  let pattern = ''
  let escaped = false
  let inClass = false
  let classStart = 0
  for (let index = 0; index < pat.source.length; index += 1) {
    const char = pat.source.charAt(index)
    if (escaped) {
      pattern += char
      escaped = false
    } else if (char === '\\') {
      pattern += char
      escaped = true
    } else if (char === '[' && !inClass) {
      pattern += char
      inClass = true
      // A leading ] after an optional ^ is a class member.
      classStart = index + 1
      if (pat.source.charAt(classStart) === '^') classStart += 1
    } else if (char === ']' && inClass && index > classStart) {
      pattern += char
      inClass = false
    } else if (char === '.' && !inClass) pattern += '[^\\n\\udc80-\\udcff]'
    else pattern += char
  }
  return new RegExp(pattern, pat.flags)
}
